src/concatenate.jl

```diff
@@ -125,16 +159,56 @@ Base.copy(concat::Concatenated) = copyto!(similar(concat), concat)
 # default falls back to replacing interface with Nothing
 # this permits specializing on typeof(dest) without ambiguities
 # Note: this needs to be defined for AbstractArray specifically to avoid ambiguities with Base.
-@inline Base.copyto!(dest::AbstractArray, concat::Concatenated) =
-  copyto!(dest, convert(Concatenated{Nothing}, concat))
+@inline function Base.copyto!(dest::AbstractArray, concat::Concatenated)
+  return copyto!(dest, convert(Concatenated{Nothing}, concat))
+end
+
+# The following is largely copied from the Base implementation of `Base.cat`, see:
+# https://github.com/JuliaLang/julia/blob/885b1cd875f101f227b345f681cc36879124d80d/base/abstractarray.jl#L1778-L1887
+_copy_or_fill!(A, inds, x) = fill!(view(A, inds...), x)
+_copy_or_fill!(A, inds, x::AbstractArray) = (A[inds...] = x)
+
+cat_size(A) = (1,)
+cat_size(A::AbstractArray) = size(A)
+cat_size(A, d) = 1
+cat_size(A::AbstractArray, d) = size(A, d)
+
+cat_indices(A, d) = Base.OneTo(1)
+cat_indices(A::AbstractArray, d) = axes(A, d)
+
+function __cat!(A, shape, catdims, X...)
+  return __cat_offset!(A, shape, catdims, ntuple(zero, length(shape)), X...)
+end
+function __cat_offset!(A, shape, catdims, offsets, x, X...)
+  # splitting the "work" on x from X... may reduce latency (fewer costly specializations)
+  newoffsets = __cat_offset1!(A, shape, catdims, offsets, x)
+  return __cat_offset!(A, shape, catdims, newoffsets, X...)
+end
+__cat_offset!(A, shape, catdims, offsets) = A
+function __cat_offset1!(A, shape, catdims, offsets, x)
+  inds = ntuple(length(offsets)) do i
+    (i <= length(catdims) && catdims[i]) ? offsets[i] .+ cat_indices(x, i) : 1:shape[i]
+  end
+  _copy_or_fill!(A, inds, x)
+  newoffsets = ntuple(length(offsets)) do i
+    (i <= length(catdims) && catdims[i]) ? offsets[i] + cat_size(x, i) : offsets[i]
+  end
+  return newoffsets
+end
+
+dims2cat(dims::Val) = dims2cat(unval(dims))
+function dims2cat(dims)
+  if any(≤(0), dims)
+    throw(ArgumentError("All cat dimensions must be positive integers, but got $dims"))
+  end
+  return ntuple(in(dims), maximum(dims))
+end
 
-# couple back to Base implementation if no specialization exists:
-# https://github.com/JuliaLang/julia/blob/29da86bb983066dd076439c2c7bc5e28dbd611bb/base/abstractarray.jl#L1852
 function Base.copyto!(dest::AbstractArray, concat::Concatenated{Nothing})
-  catdims = Base.dims2cat(dims(concat))
-  shape = Base.cat_size_shape(catdims, concat.args...)
+  catdims = dims2cat(dims(concat))
```

Comment:
I just noticed that the implementation of `copyto!(dest, concat::Concatenated{Nothing})` now got split from the one for `copyto!(dest, concat::Concatenated)` which calls the second one. Maybe we can move these together to improve clarity?

Reply:
I don't mean the logic, it's just that the specializations you added are added inbetween these two methods, which were previously together, that's all.

I do like having it here: it stems from `Base.Broadcast` as a way to allow you to dispatch on the `dest` type instead of the interface type, which would otherwise lead to ambiguities.

Reply:
It's just about the source code, the new functionality is pasted inbetween two implementations of the same function name, which is slightly strange as an organization.

The `Nothing` is there to allow you to specialize on the destination type without having method ambiguities, as otherwise `Base.copyto!(dest::MyArray, concat::Concatenated)` would be ambiguous